src/main/java/com/maddyhome/idea/vim/extension/nerdtree/Mappings.kt

```diff
@@ -0,0 +1,168 @@
+   * J........Jump down inside directories at the current tree depth.....*NERDTreeMapJumpLastChild*
+   * <C-J>....Jump down to next sibling of the current directory.......*NERDTreeMapJumpNextSibling*
+   * <C-K>....Jump up to previous sibling of the current directory.....*NERDTreeMapJumpPrevSibling*
+   * </code></pre>
+   */
+  fun registerNavigationMappings() {
+    // TODO support going [count] lines upward/downward or to line [count]
+    register("k", Action.ij("Tree-selectPrevious"))
+    register("j", Action.ij("Tree-selectNext"))
+    register("G", Action.ij("Tree-selectLast"))
+    register("gg", Action.ij("Tree-selectFirst"))
+
+    // FIXME lazy loaded tree nodes are not expanded
+    register("NERDTreeMapOpenRecursively", "O", Action.ij("FullyExpandTreeNode"))
+    // This action respects `ide.tree.collapse.recursively`. We may prompt the user to disable it
+    register("NERDTreeMapCloseDir", "x", Action.ij("Tree-selectParentNoCollapse", "CollapseTreeNode"))
+    register(
+      "NERDTreeMapCloseChildren",
+      "X",
+      // We should not rely on `ide.tree.collapse.recursively` since it closes visible paths only
+      Action { _, tree ->
+        val selectedPath = tree.selectionPath ?: return@Action
+        val expanded = tree.isExpanded(selectedPath)
+        // Collapsing children will expand their parent
+        // So we have to collapse the directories from the inside out
+        tree.nonLeafDescendants(selectedPath).reversed().forEach(tree::collapsePath)
+        // And manually collapse the current directory if applicable
+        if (!expanded) tree.collapsePath(selectedPath)
+      },
+    )
+
+    register("NERDTreeMapJumpRoot", "P", Action.ij("Tree-selectFirst"))
+    register("NERDTreeMapJumpParent", "p", Action.ij("Tree-selectParentNoCollapse"))
+    register(
+      "NERDTreeMapJumpFirstChild",
+      "K",
+      Action { _, tree ->
+        var path = tree.selectionPath ?: return@Action
+        while (true) {
+          val previous = TreeUtil.previousVisibleSibling(tree, path)
+          if (previous == null) break
+          path = previous
+        }
+        tree.selectionPath = path
+        TreeUtil.scrollToVisible(tree, path, false)
+      },
+    )
+    register(
+      "NERDTreeMapJumpLastChild",
+      "J",
+      Action { _, tree ->
+        var path = tree.selectionPath ?: return@Action
+        while (true) {
+          val next = TreeUtil.nextVisibleSibling(tree, path)
+          if (next == null) break
+          path = next
+        }
+        tree.selectionPath = path
+        TreeUtil.scrollToVisible(tree, path, false)
+      },
+    )
+    register("NERDTreeMapJumpNextSibling", "<C-J>", Action.ij("Tree-selectNextSibling"))
+    register("NERDTreeMapJumpPrevSibling", "<C-K>", Action.ij("Tree-selectPreviousSibling"))
+  }
+}
+
+private fun JTree.nonLeafDescendants(path: TreePath): List<TreePath> {
+  if (model.isLeaf(path.lastPathComponent)) return emptyList()
```

Comment:
Correct me if I'm wrong, but it looks like you're traversing the whole tree in order to close the leaf.

We should not do that, this will be a huge performance issue both mathematically and in reality.
The tree can be really huge, think about some `node_modules` folder where a half of the npm can be downloaded. Traversing the whole tree will be really slow.
Also, there is a thing called an async tree where the elements are not loaded unless they're needed. This is really useful in, for example, the project tree where you have to access the file system in order to get the tree information. Calling `getChild` will cause loading the whole tree.

Reply:
Let's adopt the IJ implementation. 
Temporarily enabling `ide.tree.collapse.recursively` sounds easier for me to use. However, try to write a test that will check that this option still exists in IJ.